File: NWMv4/scripts/identify_donors_main.R
# Create donor-receiver pairing using Principal Component Analysis (PCA) and 
#    Gower's distance, using different sets of attributes (HLR or CAMELS) 
# Multiple rounds are conducted to handle spatial gaps (NA values) in some attributes 
#
# This version allows donors and receivers to have different hydrofabric files

rm(list=ls())

library(data.table)
library(yaml)

source("apply_pca.R")
source("apply_donor_constraints.R")
source("gower_dist.R")
source("kmeans_clust.R")
source("compute_dist_spatial.R")

# Read the configuration file and assign parameters
config <- read_yaml("../data/config.yaml",readLines.warn=FALSE)

# donor attributes
dtAttrDonor <- get(load("../data/all_attrs_donors.Rdata"))
dtAttrDonor$tag <- "donor"

# receiver attributes
dtAttrReceiver <- get(load("../data/all_attrs_receivers.Rdata"))
dtAttrReceiver$tag <- "receiver"

# attributes for both donors and receivers
dtAttrAll <- rbind(dtAttrDonor, dtAttrReceiver)
dtAttrAll[,snowy:=ifelse(snow_frac>=config$pars$general$minSnowFrac,TRUE,FALSE)] #snowiness
rm(dtAttrDonor, dtAttrReceiver)

# columns in the attrs table that are not actual attributes
config$non_attr_cols <- c("id","tag","snowy","hsg")

# compute spatial distance between all receivers and donors if not already exists
f1 <- paste0("../data/dist_spatial_donor_receiver.Rdata")
if (file.exists(f1)) {
  distSpatial0 <- get(load(f1))
} else {
  distSpatial0 <- compute_dist_spatial(f1, subset(dtAttrAll,tag=="donor")$id, subset(dtAttrAll,tag=="receiver")$id)
}

# generate the donor-receiver pairing for the defined attr scenarios using different algorithms (functions)
scenarios <- names(config$attrs)
scenarios <- scenarios[scenarios!="base"]
funcs <- c("gower_dist","kmeans_clust","pamk_clust")
for (func1 in funcs) {
  for (scenario in scenarios) {
    
    outfile <- paste0("../output/donor_",scenario,"_",func1,".Rdata")
    if (file.exists(outfile)) next
    
    stop()
    message(paste0("\n======== processing donors: ", func1, "  ", scenario," =============\n"))
    dtDonorAll <- do.call(func1,list(config, dtAttrAll,scenario, distSpatial0))
    
    # if receivers and donors share the same hydrofabric, assign donors directly
    if (ver_donor == ver_receiver) {
      dtDonorAll <- subset(dtDonorAll,! id %in% donorsAll)
      dt_donor <- data.table(id=donorsAll, donor=donorsAll, distAttr=0, distSpatial=0, tag="donor")
      dtDonorAll <- rbind(dt_donor,dtDonorAll)
    }
    
    save(dtDonorAll,file=outfile)
    
  } #loop scenario
} # loop func

#dt1 <- copy(dtDonorAll)
#load("output/v0/donor_camels_kge_nse.Rdata")
#dt2 <- merge(dt1,dtDonorAll,by="id")
#dt2[,same:=ifelse(donor.x==donor.y, TRUE,FALSE)]
#subset(dt2,!same)

File: NWMv4/algorithm/identify_donors_main.py
import yaml
import pandas as pd
import os.path
import time

import funcs_clust
import funcs_dist

# read configuration (algorithm parameters etc) into dictionary
with open('../data/config.yaml', 'r') as stream:
    try:
        config = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)

# read donor attributes
dtAttrDonor = pd.read_csv('../data/all_attrs_donors.csv')
dtAttrDonor['tag'] = 'donor'

# read receiver attributes
dtAttrReceiver = pd.read_csv('../data/all_attrs_receivers.csv')
dtAttrReceiver['tag'] = 'receiver'

# if donors overlap with receivers (i.e., they share the same hydrofabric files), exclude them from the receivers
if config['hydrofabric']['shared']:
    dtAttrReceiver = dtAttrReceiver.loc[~dtAttrReceiver['id'].isin(dtAttrDonor['id'])]

# combine donor & receiver attributes
dtAttrAll = pd.concat([dtAttrDonor, dtAttrReceiver])

# detemine whether the catchment is snowy (as snowy and non-snowy catchments are processed separately)
dtAttrAll['snowy'] = dtAttrAll['snow_frac'].apply(lambda x: True if x >= config['pars']['general']['minSnowFrac'] else False)
del dtAttrDonor, dtAttrReceiver

# columns in the attrs table that are not actual attributes
config['non_attr_cols'] = ["id","tag","snowy","hsg"]

# pretty print nested dictionaries
# print(yaml.dump(config, default_flow_style=False)) 

# compute spatial distance between all receivers and donors if not already exists
f1 = '../data/dist_spatial_donor_receiver.csv'
if os.path.isfile(f1):
    distSpatial0 = pd.read_csv(f1,index_col=0)
else:
    pass
    #distSpatial0 = compute_dist_spatial(f1, subset(dtAttrAll,tag=="donor")$id, subset(dtAttrAll,tag=="receiver")$id)


# different attribute scenarios
scenarios = list(config['attrs'].keys())
scenarios.remove('base') # the base scenario is only used together with CAMELS or HLR
#scenarios = ['camels']
#scenarios = ['hlr']

# different regionalization algorithms
functions = {'urf': funcs_dist,
             'gower': funcs_dist,
             'kmeans': funcs_clust,
             'kmedoids': funcs_clust,
             'dbscan': funcs_clust,
             'hdbscan': funcs_clust,
             'birch': funcs_clust,
             }
funcs = functions.keys()
funcs = ['hdbscan']
funcs = ['kmedoids']
funcs = ['birch']

# loop through regionalization algorithms and scenarios to generate donor-receiver pairings
for func1 in funcs:
    for scenario in scenarios:
        outfile = '../output/donor_' + scenario + '_' + func1 + '.csv'

        print("\n======== processing donors: " + func1 + "  " + scenario +" =============\n")       
        start_time = time.time()
        dtDonorAll = functions[func1].func(config, dtAttrAll,scenario, distSpatial0, func1)           
        print("\nTotal processing time: --- %s seconds ---" % (time.time() - start_time))  
        
        # save donor receiver pairing to csv file    
        dtDonorAll.to_csv(outfile, index=False)

